Adding Pubmine ad network to amp-ad

dbspringer opened the pull request:
Integrates the Pubmine ad network as part of amp-ad. 

Closes https://github.com/ampproject/amphtml/issues/4471

dbspringer commented:
Hmm... not sure what the issue/solution here is. Any insight would be appreciated.

<img width="976" alt="screen shot 2016-08-18 at 12 34 33 pm" src="https://cloud.githubusercontent.com/assets/273708/17787957/36a08dc2-6540-11e6-9ec3-f0050feed393.png">

dbspringer commented:
Also, in terms of signing the CLA, should one person sign for the organization, or does each person sign for the organization and the Authorized Signer approves it?

dbspringer commented:
I signed the CLA as a corporation: Automattic Inc. Also, using my work email: derek@a8c.com.

dvoytenko commented:
/to @lannka

lannka commented:
@dbspringer the error you saw is from the type-checks of the compiler. To get around with it, you can switch to `data['sideid']`.

The PR looks good in general. Please try to sync to the latest head to address conflicts, as there was a small refactoring.

For CLA, it's still a black box to me :-(. Please try to commit your new changes (change git email) using the same email address that you signed for the CLA. But forgive me if that still doesn't work... then we'll have someone to force merge for you...

lannka commented on examples/ads.amp.html:
```diff
@@ -530,6 +531,15 @@ <h2 id="plista">Plista responsive widget</h2>
       data-categories="politik">
   </amp-embed>
 
+  <h2 id="pubmine" class="broken">Pubmine 300x250</h2>
```
remove `class="broken"`. that's for broken examples.

lannka commented on examples/ads.amp.html:
```diff
@@ -90,6 +90,7 @@
     <a href="#openadstream">Open AdStream</a> |
     <a href="#openx">OpenX</a> |
     <a href="#plista">Plista</a> |
+    <a href="#pubmine" class="broken">Pubmine</a> |
```
remove class="broken". that's for broken examples.

lannka commented:
Thanks you googlebot! 

@dbspringer what did you do to amuse it?

dbspringer commented:
@lannka Hah, good question! I updated my GitHub email for this repo, maybe that kicked it off. Otherwise, I rebased/refactored my code w/ the updates.

All other things being equal I still need a bit more time working to get the ad server-side straightened out before we merge the PR. I'm having some trouble w/ some scripts getting loaded via unsecure HTTP that I think AMP is cutting off for not using HTTPS.

dbspringer commented:
Howdy @lannka et al, I think we're good to go on our end! Please let me know what else we need to do to get the PR accepted. 

Thanks!

lannka commented on ads/pubmine.js:
```diff
@@ -0,0 +1,81 @@
+ * @param {!Object} data
+ */
+export function pubmine(global, data) {
+  validateData(data, pubmineRequired, pubmineFields);
```
the 2nd param of `validateData` is `optionalFields`. So you can rename the var and remove `'siteid'` from it.

lannka commented on ads/pubmine.js:
```diff
@@ -0,0 +1,81 @@
+  let __ATA;
+  (function(__ATA, global, data) {
+    const ipwCustom = {
+        adSafe: 'adsafe' in data ? data.adsafe : '0',
```
nitpick: indentation of line 34 - 44 is a bit off.

dbspringer replied:
Linter barks at me if I do anything different.

lannka commented:
LGTM with some minor comments.

dbspringer commented:
Thanks @lannka! I updated the variable name for the data fields. Otherwise, the linter barks at me if I try to change the indentation at all.

lannka commented:
Good to go. will merge once Travis turns green.

dbspringer commented:
Thanks @lannka! 🙇

dbspringer commented:
Howdy again @lannka, I am curious how soon the updates will be reflected in the production code. I'm trying to test the new `pubmine` amp-ad on our system and I'm getting the error `f.js:6 Uncaught Error: Unknown 3p: pubmine​​​`.

jridgewell commented:
You didn't make last weeks canary build, so you'll be in this week's (cuts are on Thursday). After a week in canary, it'll be promoted to production the following Thurs.

dbspringer commented:
@jridgewell Thanks Justin, I was just reading up on the release process--I'll give it a shot once the latest Canary goes out this week.
